packages/attestation-service/src/attestation.ts

```diff
@@ -1,38 +1,128 @@
+import { AttestationState } from '@celo/contractkit/lib/wrappers/Attestations'
 import { attestToIdentifier, SignatureUtils } from '@celo/utils'
+import { privateKeyToAddress } from '@celo/utils/lib/address'
 import { retryAsyncWithBackOff } from '@celo/utils/lib/async'
 import express from 'express'
+import * as t from 'io-ts'
+import { existingAttestationRequest, kit, persistAttestationRequest } from './db'
+import { Address, AddressType, E164Number, E164PhoneNumberType } from './request'
 import { sendSms } from './sms'
-function signAttestation(phoneNumber: string, account: string) {
+
+export const AttestationRequestType = t.type({
+  phoneNumber: E164PhoneNumberType,
+  account: AddressType,
+  issuer: AddressType,
+})
+
+export type AttestationRequest = t.TypeOf<typeof AttestationRequestType>
+
+function getAttestationKey() {
   if (process.env.ATTESTATION_KEY === undefined) {
     console.error('Did not specify ATTESTATION_KEY')
     throw new Error('Did not specify ATTESTATION_KEY')
   }
 
-  const signature = attestToIdentifier(phoneNumber, account, process.env.ATTESTATION_KEY)
+  return process.env.ATTESTATION_KEY
+}
+
+async function validateAttestationRequest(request: AttestationRequest) {
+  // check if it exists in the database
+  if (
+    (await existingAttestationRequest(request.phoneNumber, request.account, request.issuer)) !==
+    null
+  ) {
+    throw new Error('Attestation already sent')
```

Comment:
So this implies users cannot request another delivery if the SMS did not arrive?

Reply:
Correct, not at this moment, though we could consider circumstances in which the user could re-request.

Reply:
Re-requesting is a pretty useful debugging tool for verification atm. I'd say (anecdotally) half the time I get 2/3, the third comes if I retry verification. So a better alternative here would be a retry threshold, which could be just 1 time for now. But I understand that's a fair bit more complexity here so open to deferring that to another PR if you prefer